### PR TITLE
Enhancement/3374 user survey controller

### PR DESCRIPTION
**Pull request by eugene-manuilov**
## Summary

Addresses issue #3374 

## Relevant technical choices

* Fixed wrong user input settings endpoint schema.

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [ ] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

File: tests/phpunit/integration/Core/User_Surveys/REST_User_Surveys_ControllerTest.php [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+ * @link      https://sitekit.withgoogle.com
+ */
+
+namespace Google\Site_Kit\Tests\Core\User_Surveys;
+
+use Google\Site_Kit\Context;
+use Google\Site_Kit\Core\Authentication\Authentication;
+use Google\Site_Kit\Core\User_Surveys\REST_User_Surveys_Controller;
+use Google\Site_Kit\Tests\TestCase;
+
```
Comment (aaemnnosttv) [inline tests/phpunit/integration/Core/User_Surveys/REST_User_Surveys_ControllerTest.php:17]: Let's add a `User_Surveys` group similar to other core test cases.

File: includes/Core/User_Surveys/REST_User_Surveys_Controller.php [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+ * @package   Google\Site_Kit\Core\User_Surveys
+ * @copyright 2021 Google LLC
+ * @license   https://www.apache.org/licenses/LICENSE-2.0 Apache License 2.0
+ * @link      https://sitekit.withgoogle.com
+ */
+
+namespace Google\Site_Kit\Core\User_Surveys;
+
+use Google\Site_Kit\Core\Authentication\Authentication;
+use Google\Site_Kit\Core\Permissions\Permissions;
```
Comment (aaemnnosttv) [inline includes/Core/User_Surveys/REST_User_Surveys_Controller.php:14]: This import is unused.
```suggestion
```

File: tests/phpunit/integration/Core/User_Surveys/REST_User_Surveys_ControllerTest.php [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+	public function test_register() {
+		remove_all_filters( 'googlesitekit_rest_routes' );
+
+		$controller = new REST_User_Surveys_Controller( $this->authentication );
+		$controller->register();
+
+		$this->assertTrue( has_filter( 'googlesitekit_rest_routes' ) );
+	}
+
```
Comment (aaemnnosttv) [inline tests/phpunit/integration/Core/User_Surveys/REST_User_Surveys_ControllerTest.php:41]: This is a good start but let's add at least one test for each route, similar to `REST_Feature_Tours_ControllerTest`.

We don't need to worry about requests hitting the proxy since these will be blocked by `WP_HTTP_BLOCK_EXTERNAL`. We can assume that if the request responds with a `http_request_not_executed` error, then it was successfully sent to the proxy. It would be good to assert that the request was made with the expected args in the body as well.

**Review by aaemnnosttv — CHANGES_REQUESTED** [review]
Thanks @eugene-manuilov – this is off to a great start. There's one thing that should change due to an oversight in the ACs although that's rather small. The main thing left is to add a few tests for the added endpoints, but that shouldn't take too long I think.

**Comment by aaemnnosttv** [comment]
Also, please remember to add a QA Brief before sending for review so that it's ready for QA once merged 👍

**Comment by eugene-manuilov** [comment]
> Also, please remember to add a QA Brief before sending for review so that it's ready for QA once merged

Honestly, I prefer to add it after a PR is merged.

File: includes/Core/User_Surveys/REST_User_Surveys_Controller.php [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+				'core/user/data/survey-trigger',
+				array(
+					'methods'             => WP_REST_Server::CREATABLE,
+					'callback'            => function ( WP_REST_Request $request ) {
+						$proxy = $this->authentication->get_google_proxy();
+						$creds = $this->authentication->credentials();
+						$token = $this->authentication->get_oauth_client()->get_access_token();
+						$data  = $request->get_param( 'data' );
+
+						$response = $proxy->send_survey_trigger( $creds, $token, $data['token_id'] );
```
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply includes/Core/User_Surveys/REST_User_Surveys_Controller.php]: Updated.

File: includes/Core/User_Surveys/REST_User_Surveys_Controller.php [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
+						return $response;
+					},
+					'permission_callback' => $can_authenticate,
+					'args'                => array(
+						'data' => array(
+							'type'       => 'object',
+							'required'   => true,
+							'properties' => array(
+								'trigger_id' => array(
```
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply includes/Core/User_Surveys/REST_User_Surveys_Controller.php]: Updated.

File: tests/phpunit/integration/Core/User_Surveys/REST_User_Surveys_ControllerTest.php (hunk 1 above)
Comment (eugene-manuilov) [reply tests/phpunit/integration/Core/User_Surveys/REST_User_Surveys_ControllerTest.php:17]: Not sure what you mean :thinking: Which group do I need to add and where?

File: includes/Core/User_Surveys/REST_User_Surveys_Controller.php (hunk 2 above)
Comment (eugene-manuilov) [reply includes/Core/User_Surveys/REST_User_Surveys_Controller.php]: Removed.

File: tests/phpunit/integration/Core/User_Surveys/REST_User_Surveys_ControllerTest.php (hunk 3 above)
Comment (eugene-manuilov) [reply tests/phpunit/integration/Core/User_Surveys/REST_User_Surveys_ControllerTest.php:41]: I have added tests for Google_Proxy methods that test requests to the proxy. It should be enough I guess, shouldn't it?

File: tests/phpunit/integration/Core/User_Surveys/REST_User_Surveys_ControllerTest.php (hunk 1 above)
Comment (aaemnnosttv) [reply tests/phpunit/integration/Core/User_Surveys/REST_User_Surveys_ControllerTest.php:17]: Ah sorry, I was referring to a PHPUnit group annotation: `@group User_Surveys` to the test's doc block.

Looking again it's probably not necessary right now since there probably wouldn't be anything else to go in it, so let's not worry about it for now.

File: tests/phpunit/integration/Core/User_Surveys/REST_User_Surveys_ControllerTest.php (hunk 3 above)
Comment (aaemnnosttv) [reply tests/phpunit/integration/Core/User_Surveys/REST_User_Surveys_ControllerTest.php:41]: Thank you for adding those for `Google_Proxy` 👍 

I would prefer if we had tests that actually invoke the registered routes as more of an end-to-end test although we wouldn't have to duplicate everything that's covered by the other tests you added.

If you think this would take too much time we can skip it in this case but I think we should try to add better coverage for our REST routes going forward.

**Review by aaemnnosttv — COMMENTED** [review]
This is good to go regarding what is here. One additional thought about the tests for our routes for you: https://github.com/google/site-kit-wp/pull/3440#discussion_r644869094

File: tests/phpunit/integration/Core/User_Surveys/REST_User_Surveys_ControllerTest.php (hunk 3 above)
Comment (eugene-manuilov) [reply tests/phpunit/integration/Core/User_Surveys/REST_User_Surveys_ControllerTest.php:41]: Added a test that checks the registered route arguments. Testing REST route handlers themself is quite cumbersome and will take too much time to do it properly. Let's skip it.

**Review by aaemnnosttv — APPROVED** [review]
Great, thanks @eugene-manuilov !
